tests/testthat/test-derive_locf_records.R

```diff
@@ -0,0 +1,158 @@
+    ~PARAMCD, ~PARAM,                           ~AVISITN, ~AVISIT,
+    "DIABP", "Diastolic Blood Pressure (mmHg)", 0,       "BASELINE",
+    "DIABP", "Diastolic Blood Pressure (mmHg)", 2,       "WEEK 2",
+    "SYSBP", "Systolic Blood Pressure (mmHg)",  0,       "BASELINE",
+    "SYSBP", "Systolic Blood Pressure (mmHg)",  2,       "WEEK 2",
+  )
+
+
+  expected_output <- bind_rows(
+    input,
+    tibble::tribble(
+      ~STUDYID,       ~USUBJID,     ~PARAMCD, ~PARAM,                           ~AVAL, ~AVISITN, ~AVISIT,
+      "CDISCPILOT01", "01-701-1028", "DIABP", "Diastolic Blood Pressure (mmHg)", 79,    2,        "WEEK 2",
+      "CDISCPILOT01", "01-701-1028", "SYSBP", "Systolic Blood Pressure (mmHg)", 130,    2,        "WEEK 2"
+    ) %>%
+      mutate(DTYPE = "LOCF"))
+
+
+  actual_output <- derive_locf_records(
+    input,
+    dataset_expected_obs=advs_expected_obsv,
+    by_vars=vars(STUDYID, USUBJID, PARAM, PARAMCD),
+    order=vars(STUDYID, USUBJID, PARAM, PARAMCD, AVISITN, AVISIT))
+
+
+  expect_dfs_equal(
+    base = expected_output,
+    compare = actual_output,
+    keys = c("STUDYID", "USUBJID", "PARAMCD", "PARAM", "AVISITN", "AVISIT", "DTYPE")
+  )
+
+})
+
+
+test_that("derive_locf_records Test 3: new locf observations are derived correctly when visits are missing - and DTYPE already exits", {
+
+
+
+  input <- tibble::tribble(
+    ~STUDYID,       ~USUBJID,     ~PARAMCD,   ~PARAM,                                 ~AVAL, ~AVISITN, ~AVISIT,    ~DTYPE,
+    "CDISCPILOT01", "01-701-1015", "DIABP",   "Diastolic Blood Pressure (mmHg)",      51,    0,        "BASELINE", NA,
+    "CDISCPILOT01", "01-701-1015", "DIABP",   "Diastolic Blood Pressure (mmHg)",      50,    2,        "WEEK 2",   NA,
+    "CDISCPILOT01", "01-701-1015", "SYSBP",   "Systolic Blood Pressure (mmHg)",       121,   0,        "BASELINE", NA,
+    "CDISCPILOT01", "01-701-1015", "SYSBP",   "Systolic Blood Pressure (mmHg)",       121,   2,        "WEEK 2",   NA,
+    "CDISCPILOT01", "01-701-1015", "LTDIABP", "Log(Diastolic Blood Pressure (mmHg))", 1.71,  0,        "BASELINE", 'LOG',
+    "CDISCPILOT01", "01-701-1015", "LTDIABP", "Log(Diastolic Blood Pressure (mmHg))", 1.69,  2,        "WEEK 2",   'LOG',
+    "CDISCPILOT01", "01-701-1015", "LTSYSBP", "Log(Systolic Blood Pressure (mmHg))",  2.08,  0,        "BASELINE", 'LOG',
+    "CDISCPILOT01", "01-701-1015", "LTSYSBP", "Log(Systolic Blood Pressure (mmHg))",  2.08,  2,        "WEEK 2",   'LOG',
+    "CDISCPILOT01", "01-701-1028", "DIABP",   "Diastolic Blood Pressure (mmHg)",      79,    0,        "BASELINE", NA,
+    "CDISCPILOT01", "01-701-1028", "SYSBP",   "Systolic Blood Pressure (mmHg)",      130,   0,        "BASELINE", NA,
+    "CDISCPILOT01", "01-701-1028", "LTDIABP", "Log(Diastolic Blood Pressure (mmHg))", 1.89,  0,        "BASELINE", 'LOG',
+    "CDISCPILOT01", "01-701-1028", "LTSYSBP", "Log(Systolic Blood Pressure (mmHg))",  2.11,  0,        "BASELINE", 'LOG'
+  )
+
+  advs_expected_obsv <- tibble::tribble(
```

Comment:
Could you make the visits parameter dependent in one of the test cases?

Reply:
Added this case as 'Test 5'.